### PR TITLE
Refactor

### PR DESCRIPTION
**Pull request by hachilam92**
Achieved:
1. adjust the file structure of interfaces, group them with the corresponding services;
2. move the business logic and data manipulations in AccountController and UserController to AccountService and UserService;
3. move the Cors origin to the configuration file;
4. replace the ExceptionMiddleware with an ExceptionFilter;

### CODE REVIEW

File: API/Filters/ExceptionFilter.cs [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+{
+	public class ExceptionFilter : IAsyncExceptionFilter
+	{
+		public async Task OnExceptionAsync(ExceptionContext context)
+		{
+			Console.WriteLine(context.Exception.Message);
+			
+			var response = new ApiException(context.HttpContext.Response.StatusCode, "Internal server error");
+			
+			var options = new JsonSerializerOptions
```
Comment (aussieweever) [inline API/Filters/ExceptionFilter.cs:18]: Generally the JsonSerializerOptions should be aligned in the whole soultion, so suggest to move this options setting to the Startup.cs as a global setting

File: API/Controllers/UserController.cs [hunk 2]
```diff
@@ -11,81 +11,63 @@
...
+            if (isUpdated) { return NoContent(); }
 
             return BadRequest("Failed to update user");
         }
 
         [HttpPost("add-photo")]
         public async Task<ActionResult<PhotoDTO>> AddPhoto(IFormFile file)
         {
-            var user = await _userRepository.GetUserByUsernameAsync(User.GetUsername());
+            var user = await _userService.GetUserByUsername(User.GetUsername());
```
Comment (aussieweever) [inline API/Controllers/UserController.cs:62]: Good to see you wrap the userRepository into the userService, and you can also inject the PhotoService into the UserService. As generally, we don't have the business logic in the controller

File: API/Services/PhotoService.cs [hunk 3]
```diff
@@ -1,18 +1,21 @@
...
 {
   public class PhotoService : IPhotoService
   {
     private readonly Cloudinary _cloudinary;
+    private readonly IUserRepository _userRepository;
 
-    public PhotoService(IOptions<CloudinarySettings> config)
+    public PhotoService(IOptions<CloudinarySettings> config, IUserRepository userRepository)
     {
         var acc = new Account
```
Comment (aussieweever) [inline API/Services/PhotoService.cs:20]: Not sure the logic here, just use a default Account without any values populated?

File: API/Services/PhotoService.cs [hunk 4]
```diff
@@ -22,8 +25,27 @@ public PhotoService(IOptions<CloudinarySettings> config)
         );
 
         _cloudinary = new Cloudinary(acc);
```
Comment (aussieweever) [inline API/Services/PhotoService.cs:27]: Possible to register the Cloudinary in Startup.cs and then reject it into the PhotoService?

File: API/Services/PhotoService.cs [hunk 5]
```diff
@@ -22,8 +25,27 @@ public PhotoService(IOptions<CloudinarySettings> config)
...
+
+    public async Task<Photo> AddPhotoAsync(ImageUploadResult result, AppUser user)
+    {
+		var photo = new Photo
+		{
+			Url = result.SecureUrl.AbsoluteUri,
+			PublicId = result.PublicId
+		};
+
+		if(user.Photos.Count == 0)
```
Comment (aussieweever) [inline API/Services/PhotoService.cs:39]: Can be simplified to photo.IsMain = user.Photos.Any();

File: API/Startup.cs [hunk 6]
```diff
@@ -49,7 +53,7 @@ public void Configure(IApplicationBuilder app, IWebHostEnvironment env)
 
             app.UseRouting();
 
-            app.UseCors(X => X.AllowAnyHeader().AllowAnyMethod().WithOrigins("https://localhost:4200"));
+            app.UseCors(X => X.AllowAnyHeader().AllowAnyMethod().WithOrigins(_config["Origin"]));
```
Comment (aussieweever) [inline API/Startup.cs:56]: Origins should be an array, and better to configure the settings to make it as an array

File: API/Filters/ExceptionFilter.cs (hunk 1 above)
Comment (hachilam92) [reply API/Filters/ExceptionFilter.cs:18]: I tried to globally set the JSON option in startup.cs like:
            services
                .AddControllers(_config => {
                    _config.Filters.Add(new ExceptionFilter());
                })
                .AddJsonOptions(option => {
                    option.JsonSerializerOptions.PropertyNamingPolicy = JsonNamingPolicy.CamelCase;
                });

However, it didn't work, is there any mistake I made here?

**Pull request merged**